satpy/composites/__init__.py

```diff
@@ -770,60 +770,104 @@ class DayNightCompositor(GenericCompositor):
-            coszen = np.cos(np.deg2rad(projectables[2]))
-
-            coszen -= min(np.cos(np.deg2rad(lim_high)),
-                          np.cos(np.deg2rad(lim_low)))
-            coszen /= np.abs(np.cos(np.deg2rad(lim_low)) -
-                             np.cos(np.deg2rad(lim_high)))
-            coszen = np.clip(coszen, 0, 1)
-
-            full_data = []
-
-            # Apply enhancements
-            day_data = enhance2dataset(day_data)
-            night_data = enhance2dataset(night_data)
-
-            # Match dimensions to the data with more channels
-            # There are only 1-channel and 3-channel composites
-            if day_data.shape[0] > night_data.shape[0]:
-                night_data = np.ma.repeat(night_data, 3, 0)
-            elif day_data.shape[0] < night_data.shape[0]:
-                day_data = np.ma.repeat(day_data, 3, 0)
-
-            for i in range(day_data.shape[0]):
-                day = day_data[i, :, :]
-                night = night_data[i, :, :]
-
-                data = (1 - coszen) * np.ma.masked_invalid(night).filled(0) + \
-                    coszen * np.ma.masked_invalid(day).filled(0)
-                data = np.ma.array(data, mask=np.logical_and(night.mask,
-                                                             day.mask),
-                                   copy=False)
-                data = Dataset(np.ma.masked_invalid(data),
-                               copy=True,
-                               **projectables[0].info)
-                full_data.append(data)
-
-            res = super(DayNightCompositor, self).__call__((full_data[0],
-                                                            full_data[1],
-                                                            full_data[2]),
-                                                           *args, **kwargs)
+    def __init__(self, lim_low=85., lim_high=95., *args, **kwargs):
+        """Collect custom configuration values.
 
-        except ValueError:
-            raise IncompatibleAreas
+        Args:
+            lim_low (float): lower limit of Sun zenith angle for the
+                             blending of the given channels
+            lim_high (float): upper limit of Sun zenith angle for the
+                             blending of the given channels
+        """
+        self.lim_low = lim_low
+        self.lim_high = lim_high
+        super(DayNightCompositor, self).__init__(*args, **kwargs)
+
+    def __call__(self, projectables, *args, **kwargs):
+
+        day_data = projectables[0]
+        night_data = projectables[1]
+
+        lim_low = np.cos(np.deg2rad(self.lim_low))
+        lim_high = np.cos(np.deg2rad(self.lim_high))
+        try:
+            coszen = xu.cos(xu.deg2rad(projectables[2]))
+        except IndexError:
+            from pyorbital.astronomy import cos_zen
+            LOG.debug("Computing sun zenith angles.")
+            lons, lats = day_data.attrs["area"].get_lonlats_dask(CHUNK_SIZE)
```

Comment:
You should probably pass `day_data.chunks` as the chunk size for potentially better performance (assuming day_data is 2D).

Reply:
I tried that, but day_data is usually, but not always, a 3D dataset, so the chunking doesn't necessarily match. I'll take a closer look when I'm back at office.

Reply:
Oh ok. Makes sense. You *might* be able to get away with `day_data.sel(bands=day_data['bands'][0]).chunks`...maybe.

Reply:
Put another way, what if you got one 2D array from the 3D array and use the chunks from that. Dask *might* give you logical chunks. The point is if you can pass the actual chunks then calculations should be overall faster since the lons/lats will be at the same chunk size as the data.